File: wrappers/swift/Tests/PrinterDriverTests/DetectionTests.swift
import CPrinterDriver
import CPrinterDriverTestSupport
import Foundation
import XCTest

@testable import PrinterDriver

/// M15 — self-test, auto-detection and LAN discovery through the Swift wrapper
/// (docs/api.md §15).
///
/// The wrapper contains no detection logic, so what is under test here is that none of it
/// is *lost* on the way across: that a Done at grade A is still a Done at grade A, that a
/// printless probe's refusal to promote a provenance survives the bridge, and that a
/// sweep which must not print still does not print.

/// A thread-safe sink for the callbacks the sweep fires from its workers.
private final class DeviceCollector: @unchecked Sendable {
  private let lock = NSLock()
  private var devices: [DiscoveredDevice] = []

  func append(_ device: DiscoveredDevice) {
    lock.lock()
    devices.append(device)
    lock.unlock()
  }

  var count: Int {
    lock.lock()
    defer { lock.unlock() }
    return devices.count
  }
}

/// M15 — self-test, auto-detection and LAN discovery through the Swift wrapper
/// (docs/api.md §15).
final class DetectionTests: XCTestCase {

  func testSelfTestPrintsOneTicketAndReportsWhatItProved() throws {
    let driver = try PrinterDriver(fsyncDisabled: true)
    let bench = try driver.scriptedPrinter(id: "bench", .healthy)

    let result: SelfTestResult = try bench.selfTest()

    // The proof is the ordinary tri-state result of the ordinary engine.
    guard case .done(let confidence, let grade, let authority, let method) = result.result
    else {
      return XCTFail("expected a Done self-test, got \(result.result)")
    }
    XCTAssertEqual(confidence, .cutFaultFree)
    XCTAssertEqual(grade, .aJobLevelConfirmation)
    XCTAssertEqual(authority, .physicalPrinter)
    XCTAssertEqual(method, "GS(H) fn48")

    XCTAssertTrue(result.key.hasPrefix("selftest-"))
    XCTAssertEqual(result.verificationID?.count, 4)
    XCTAssertNotNil(result.job)
    XCTAssertEqual(result.job?.key, result.key)

    // The ticket, and the bytes the device actually received.
    XCTAssertTrue(result.ticketLines.contains { $0.contains("PRINTERDRIVER SELF-TEST") })
    XCTAssertTrue(result.ticketLines.contains { $0.contains("CHARSET") })
    XCTAssertTrue(bench.scriptedReceivedContains("PRINTERDRIVER SELF-TEST"))
    XCTAssertTrue(bench.scriptedReceivedContains("V:"))
    XCTAssertEqual(bench.scriptedCuts, 1)

    // The detection report the paper carries.
    XCTAssertEqual(result.detection.completion.mechanism, .gsParenH)
    XCTAssertEqual(result.detection.completion.gradeCeiling, .aJobLevelConfirmation)
    XCTAssertEqual(result.detection.media.printableWidthDots, 576)
    XCTAssertEqual(result.detection.media.charactersPerLine, 48)
    XCTAssertEqual(result.detection.endpoint, "bench")
    XCTAssertTrue(result.detection.degradations.isEmpty)
    XCTAssertTrue(result.detection.provenanceSummary.contains("GS(H) fn48"))
    XCTAssertEqual(result.detection.selection.abiName, "Documented")

    // A self-test is an ordinary job under an ordinary key: the same key twice prints
    // once.
    let again = try bench.selfTest(SelfTestOptions(key: "selftest-fixed"))
    XCTAssertEqual(bench.scriptedCuts, 2)
    let third = try bench.selfTest(SelfTestOptions(key: "selftest-fixed"))
    XCTAssertEqual(bench.scriptedCuts, 2)
    XCTAssertTrue(again.job === third.job)
  }

  func testAutoDetectClassifiesLoopbackListenersAndPrintsNothing() throws {
    let answering = try XCTUnwrap(pd_test_listener_start("ok"))
    let silent = try XCTUnwrap(pd_test_listener_start("silent"))
    let gone = try XCTUnwrap(pd_test_listener_start("ok"))
    defer {
      pd_test_listener_destroy(answering)
      pd_test_listener_destroy(silent)
      pd_test_listener_destroy(gone)
    }
    let refusedPort = pd_test_listener_port(gone)
    pd_test_listener_stop(gone)  // the port is now closed: a refusal, deterministically

    let driver = try PrinterDriver(fsyncDisabled: true)
    let options = AutoDetectOptions(
      endpoints: [
        "127.0.0.1:\(pd_test_listener_port(answering))",
        "127.0.0.1:\(pd_test_listener_port(silent))",
        "127.0.0.1:\(refusedPort)",
      ],
      connectTimeoutMilliseconds: 500,
      responseTimeoutMilliseconds: 150)

    let found = try driver.autoDetect(options)
    XCTAssertEqual(found.count, 3)

    let talker = try XCTUnwrap(found.first { $0.status == .answered })
    XCTAssertTrue(talker.isPortOpen)
    XCTAssertFalse(talker.isFromCache)
    XCTAssertEqual(talker.summary.identity.model, "TM-T88V")
    // GS I is a string the firmware chooses, and at least one family ships answering as
    // somebody else's model.
    XCTAssertFalse(talker.summary.identity.isTrusted)
    XCTAssertEqual(talker.summary.completion.mechanism, .gsParenH)
    XCTAssertEqual(talker.summary.completion.gradeCeiling, .aJobLevelConfirmation)
    // The printless probe promotes the flag and not its provenance.
    XCTAssertEqual(talker.summary.completion.provenance, .unverified)
    XCTAssertTrue(talker.summary.degradations.contains { $0.contains("empty buffer") })
    XCTAssertFalse(talker.dleEotHex.isEmpty)

    let quiet = try XCTUnwrap(found.first { $0.status == .silent })
    XCTAssertTrue(quiet.isPortOpen)
    XCTAssertTrue(quiet.dleEotHex.isEmpty)

    let dead = try XCTUnwrap(found.first { $0.status == .unreachable })
    XCTAssertFalse(dead.isPortOpen)
    XCTAssertTrue(dead.summary.profileID.isEmpty)

    // The whole point: not one printable byte reached either live device.
    XCTAssertEqual(pd_test_listener_print_data_bytes(answering), 0)
    XCTAssertEqual(pd_test_listener_print_data_bytes(silent), 0)

    pd_test_listener_stop(answering)
    pd_test_listener_stop(silent)
  }

  func testAutoDetectStreamDeliversCandidatesAsTheyAreFound() throws {
    let answering = try XCTUnwrap(pd_test_listener_start("ok"))
    defer { pd_test_listener_destroy(answering) }
    let driver = try PrinterDriver(fsyncDisabled: true)
    let options = AutoDetectOptions(
      endpoints: ["127.0.0.1:\(pd_test_listener_port(answering))"],
      connectTimeoutMilliseconds: 500,
      responseTimeoutMilliseconds: 150)

    runAsync(in: self) {
      var seen: [DetectedPrinter] = []
      for try await candidate in driver.autoDetectStream(options) {
        seen.append(candidate)
      }
      XCTAssertEqual(seen.count, 1)
      XCTAssertEqual(seen.first?.status, .answered)
    }
    XCTAssertEqual(pd_test_listener_print_data_bytes(answering), 0)
    pd_test_listener_stop(answering)
  }

  func testDiscoverSweepsALoopbackAddressAndWritesOnlyDleEot() throws {
    let answering = try XCTUnwrap(pd_test_listener_start("ok"))
    defer { pd_test_listener_destroy(answering) }
    let driver = try PrinterDriver(fsyncDisabled: true)

    let options = DiscoverOptions(
      subnetCIDR: "127.0.0.1/32",
      port: pd_test_listener_port(answering),
      connectTimeoutMilliseconds: 500,
      responseTimeoutMilliseconds: 300)

    // The per-device callback runs on a sweep worker thread, so the collector is a
    // reference type with its own lock rather than a captured `var`.
    let streamed = DeviceCollector()
    let found = try driver.discover(options) { streamed.append($0) }
    XCTAssertEqual(found.count, 1)
    XCTAssertEqual(streamed.count, 1)
    XCTAssertEqual(found.first?.ip, "127.0.0.1")
    XCTAssertTrue(found.first?.isPortOpen == true)
    XCTAssertTrue(found.first?.didAnswer == true)
    // The scripted device's DLE EOT 1 answer: online, drawer pin high.
    XCTAssertEqual(found.first?.dleEotHex, "16")
    XCTAssertEqual(pd_test_listener_print_data_bytes(answering), 0)

    // A CIDR wider than /16 is a mistyped subnet, not a venue, and is refused.
    XCTAssertThrowsError(try driver.discover(DiscoverOptions(subnetCIDR: "10.0.0.0/8")))

    pd_test_listener_stop(answering)
  }

  func testTheDetectionEnumsMirrorTheAbiSpellings() {
    XCTAssertEqual(ProfileSelection.allCases.count, Int(PD_PROFILE_SELECTION_COUNT.rawValue))
    XCTAssertEqual(DetectionStatus.allCases.count, Int(PD_DETECTION_STATUS_COUNT.rawValue))
    for value in ProfileSelection.allCases {
      XCTAssertEqual(
        value.abiName,
        String(cString: pd_test_cpp_enum_name(PD_TEST_ENUM_PROFILE_SELECTION, Int32(value.rawValue))))
    }
    for value in DetectionStatus.allCases {
      XCTAssertEqual(
        value.abiName,
        String(cString: pd_test_cpp_enum_name(PD_TEST_ENUM_DETECTION_STATUS, Int32(value.rawValue))))
    }
  }
}

File: wrappers/dotnet/PrinterDriver.Tests/DetectionTests.cs
namespace PrinterDriver.Tests;

/// <summary>
/// M15 — self-test, auto-detection and LAN discovery through the managed wrapper
/// (docs/api.md §15).
/// </summary>
/// <remarks>
/// The wrapper holds no detection logic, so what is under test here is that none of it is
/// lost on the way across: that a Done at grade A is still a Done at grade A, that a
/// printless probe's refusal to promote a provenance survives the bridge, and that a sweep
/// which must not print still does not print.
/// </remarks>
public sealed class DetectionTests
{
    [Fact]
    public void SelfTestPrintsOneTicketAndReportsWhatItProved()
    {
        using var driver = TestDriver.Open();
        var bench = driver.AddScripted("ok", "bench");

        var result = bench.SelfTest();

        // The proof is the ordinary tri-state result of the ordinary engine.
        var done = Assert.IsType<JobResult.Done>(result.Result);
        Assert.Equal(ConfidenceLevel.CutFaultFree, done.Confidence);
        Assert.Equal(ConfidenceGrade.AJobLevelConfirmation, done.Grade);
        Assert.Equal(CompletionAuthority.PhysicalPrinter, done.Authority);
        Assert.Equal("GS(H) fn48", done.Method);

        Assert.StartsWith("selftest-", result.Key, StringComparison.Ordinal);
        Assert.NotNull(result.VerificationId);
        Assert.Equal(4, result.VerificationId!.Length);
        Assert.NotNull(result.Job);
        Assert.Contains(result.TicketLines, l => l.Contains("PRINTERDRIVER SELF-TEST", StringComparison.Ordinal));
        Assert.Contains(result.TicketLines, l => l.Contains("CHARSET", StringComparison.Ordinal));

        // The bytes the device actually received.
        Assert.Equal(1, TestNative.pd_test_received_contains(bench.Handle, "PRINTERDRIVER SELF-TEST"));
        Assert.Equal(1, TestNative.pd_test_received_contains(bench.Handle, "V:"));
        Assert.Equal((nuint)1, TestNative.pd_test_cuts(bench.Handle));

        // The detection report the paper carries.
        Assert.Equal("bench", result.Detection.Endpoint);
        Assert.Equal(CompletionMechanism.GsParenH, result.Detection.Completion.Mechanism);
        Assert.Equal(ConfidenceGrade.AJobLevelConfirmation, result.Detection.Completion.GradeCeiling);
        Assert.Equal(576u, result.Detection.Media.PrintableWidthDots);
        Assert.Equal(48u, result.Detection.Media.CharsPerLine);
        Assert.Empty(result.Detection.Degradations);
        Assert.Contains("GS(H) fn48", result.Detection.ProvenanceSummary, StringComparison.Ordinal);
        Assert.Equal(ProfileSelection.Documented, result.Detection.Selection);

        // A self-test is an ordinary job under an ordinary key: the same key twice
        // prints once.
        var again = bench.SelfTest(new SelfTestOptions { Key = "selftest-fixed" });
        Assert.Equal((nuint)2, TestNative.pd_test_cuts(bench.Handle));
        var third = bench.SelfTest(new SelfTestOptions { Key = "selftest-fixed" });
        Assert.Equal((nuint)2, TestNative.pd_test_cuts(bench.Handle));
        Assert.Same(again.Job, third.Job);
    }

    [Fact]
    public void AutoDetectClassifiesLoopbackListenersAndPrintsNothing()
    {
        using var answering = new ScriptedListener("ok");
        using var silent = new ScriptedListener("silent");
        using var gone = new ScriptedListener("ok");
        var refused = gone.Endpoint;
        gone.Stop(); // the port is now closed: a refusal, deterministically

        using var driver = TestDriver.Open();
        var options = new AutoDetectOptions
        {
            Endpoints = new[] { answering.Endpoint, silent.Endpoint, refused },
            ConnectTimeoutMs = 500,
            ResponseTimeoutMs = 150,
        };

        var streamed = new List<DetectedPrinter>();
        var found = driver.AutoDetect(options, streamed.Add);
        Assert.Equal(3, found.Count);
        Assert.Equal(3, streamed.Count);

        var talker = Assert.Single(found, p => p.Status == DetectionStatus.Answered);
        Assert.True(talker.IsPortOpen);
        Assert.False(talker.IsFromCache);
        Assert.Equal("TM-T88V", talker.Summary.Identity.Model);
        // GS I is a string the firmware chooses, and at least one family ships answering
        // as somebody else's model.
        Assert.False(talker.Summary.Identity.IsTrusted);
        Assert.Equal(CompletionMechanism.GsParenH, talker.Summary.Completion.Mechanism);
        Assert.Equal(ConfidenceGrade.AJobLevelConfirmation, talker.Summary.Completion.GradeCeiling);
        // The printless probe promotes the flag and not its provenance.
        Assert.Equal(Provenance.Unverified, talker.Summary.Completion.Provenance);
        Assert.Contains(talker.Summary.Degradations, l => l.Contains("empty buffer", StringComparison.Ordinal));
        Assert.NotEmpty(talker.DleEotHex);

        var quiet = Assert.Single(found, p => p.Status == DetectionStatus.Silent);
        Assert.True(quiet.IsPortOpen);
        Assert.Empty(quiet.DleEotHex);

        var dead = Assert.Single(found, p => p.Status == DetectionStatus.Unreachable);
        Assert.False(dead.IsPortOpen);
        Assert.Empty(dead.Summary.ProfileId);

        // The whole point: not one printable byte reached either live device.
        Assert.Equal((nuint)0, answering.PrintDataBytes);
        Assert.Equal((nuint)0, silent.PrintDataBytes);

        answering.Stop();
        silent.Stop();
    }

    [Fact]
    public async Task AutoDetectAsyncDeliversCandidatesAsTheyAreFound()
    {
        using var answering = new ScriptedListener("ok");
        using var driver = TestDriver.Open();
        var options = new AutoDetectOptions
        {
            Endpoints = new[] { answering.Endpoint },
            ConnectTimeoutMs = 500,
            ResponseTimeoutMs = 150,
        };

        var seen = new List<DetectedPrinter>();
        await foreach (var candidate in driver.AutoDetectAsync(options))
        {
            seen.Add(candidate);
        }
        Assert.Single(seen);
        Assert.Equal(DetectionStatus.Answered, seen[0].Status);
        Assert.Equal((nuint)0, answering.PrintDataBytes);
        answering.Stop();
    }

    [Fact]
    public void DiscoverSweepsALoopbackAddressAndWritesOnlyDleEot()
    {
        using var answering = new ScriptedListener("ok");
        using var driver = TestDriver.Open();

        var found = driver.Discover(new DiscoverOptions
        {
            SubnetCidr = "127.0.0.1/32",
            Port = answering.Port,
            ConnectTimeoutMs = 500,
            ResponseTimeoutMs = 300,
        });

        var device = Assert.Single(found);
        Assert.Equal("127.0.0.1", device.Ip);
        Assert.True(device.IsPortOpen);
        Assert.True(device.DidAnswer);
        // The scripted device's DLE EOT 1 answer: online, drawer pin high.
        Assert.Equal("16", device.DleEotHex);
        Assert.Equal((nuint)0, answering.PrintDataBytes);

        // A CIDR wider than /16 is a mistyped subnet, not a venue, and is refused.
        Assert.Throws<PrinterDriverException>(
            () => driver.Discover(new DiscoverOptions { SubnetCidr = "10.0.0.0/8" }));

        // And the local subnet is either a CIDR or an honest null.
        _ = driver.LocalSubnet;

        answering.Stop();
    }

    [Fact]
    public void TheDetectionEnumsMirrorTheCoreSpellings()
    {
        NativeFixture.Bind();
        Assert.Equal(
            Enum.GetValues<ProfileSelection>().Length,
            TestNative.pd_test_cpp_enum_count((int)BridgedEnum.ProfileSelection));
        Assert.Equal(
            Enum.GetValues<DetectionStatus>().Length,
            TestNative.pd_test_cpp_enum_count((int)BridgedEnum.DetectionStatus));

        foreach (var value in Enum.GetValues<ProfileSelection>())
        {
            Assert.Equal(
                value.ToString(),
                TestNative.ReadUtf8(TestNative.pd_test_cpp_enum_name(
                    (int)BridgedEnum.ProfileSelection, (int)value)));
        }
        foreach (var value in Enum.GetValues<DetectionStatus>())
        {
            Assert.Equal(
                value.ToString(),
                TestNative.ReadUtf8(TestNative.pd_test_cpp_enum_name(
                    (int)BridgedEnum.DetectionStatus, (int)value)));
        }
    }
}
